docs/vhlo.md

```diff
@@ -0,0 +1,128 @@
+# The VHLO Dialect
+
+## What is the VHLO Dialect?
+
+The VHLO (Versioned StableHLO) Dialect is used for serialization and stability.
+It provides a snapshot of the StableHLO dialect at a given point in time,
+denoted by version number.
+
+VHLO is an **add-only dialect** with **versioned ops**, which means that once an
+op is added to the dialect, it cannot be modified in any way that impact the
+ops semantics.
+
+Any changes to an op require a new version of an op to be added to the dialect.
+For example, if a hypothetical `my_op` was added to StableHLO in 0.9.0, but
+was changed in 0.11.0, we would have the following in VHLO:
+
+```tablegen
+// This represents the StableHLO version of the op from 0.9.0 -> 0.10.0
+def VHLO_MyOpV1 : VHLO_Op<"my_op", "0.9.0", "0.10.0"> {
+  let arguments = (ins
+    VHLO_AnyType:$operand
+  );
+  let results = (outs VHLO_AnyType:$result);
+}
+
+// This represents the StableHLO version of the op from 0.11.0 -> current
+def VHLO_MyOpV2 : VHLO_Op<"my_op_v2", "0.11.0", "current"> {
+  let arguments = (ins
+    VHLO_AnyType:$operand,
+    VHLO_AnyAttr:$attr  // New attribute added to StableHLO in 0.11.0
+  );
+  let results = (outs VHLO_AnyType:$result);
+}
+```
+
+Note that under this schema, the latest version of an op directly represents
+the StableHLO equivalent of the op.
+
+## Why is this necessary?
+
+Having an add-only snapshot dialect allows us to target previous versions of
+the StableHLO opset. This reduces forward and backward compatibility to MLIR
+conversions between ops in the VHLO dialect.
+
+**Forward compatibility:** If every op in a program can be downgraded to the
+target version, it is guaranteed to be deserializable and convertable to
+StableHLO on a consumer running a version <= the target version, since VHLO
+has a snapshot of the opset at that time.
+
+![Forward compatibility image](images/vhlo/forward_compatibility.png)
```

Comment:
I think that we should say `vhlo.f32` instead of just `f32`. Looks like there's sufficient space everywhere to accommodate that?